File: app/services/project_service.py
from sqlalchemy import select
from sqlalchemy.orm import Session

from app.models import PageStatus, Project
from app.schemas.project import ProjectCreate, ProjectUpdate


def list_projects(db: Session) -> list[Project]:
    return list(db.scalars(select(Project).order_by(Project.updated_at.desc())))


def get_project(db: Session, project_id: str) -> Project | None:
    return db.get(Project, project_id)


def create_project(db: Session, data: ProjectCreate) -> Project:
    project = Project(**data.model_dump())
    db.add(project)
    db.commit()
    db.refresh(project)
    return project


def update_project(db: Session, project: Project, data: ProjectUpdate) -> Project:
    for key, value in data.model_dump().items():
        setattr(project, key, value)
    db.commit()
    db.refresh(project)
    return project


def delete_project(db: Session, project: Project) -> None:
    db.delete(project)
    db.commit()


def project_stats(project: Project) -> dict:
    pages = project.pages
    return {
        "total": len(pages),
        "processed": sum(bool(p.processed_image_path) for p in pages),
        "transcribed": sum(bool(p.raw_ocr_text) for p in pages),
        "reviewed": sum(p.status == PageStatus.REVIEWED for p in pages),
        "needs_review": sum(p.status == PageStatus.NEEDS_REVIEW for p in pages),
    }


def missing_page_numbers(project: Project) -> list[int]:
    numbers = sorted({p.page_number for p in project.pages})
    if len(numbers) < 2:
        return []
    return [n for n in range(numbers[0], numbers[-1] + 1) if n not in numbers]

File: src/services/project-service.js
import { getProject, listPages, now, saveProject, uid } from './database.js'

export function newProject(data) {
  const timestamp = now()
  return {
    id: uid(), title: data.title.trim(), author: data.author?.trim() || '',
    description: data.description?.trim() || '', language: data.language || 'por',
    transcriptionMode: data.transcriptionMode || 'diplomatic',
    includeHandwrittenNotes: Boolean(data.includeHandwrittenNotes),
    status: 'active', createdAt: timestamp, updatedAt: timestamp,
  }
}

export async function projectWithPages(id) {
  const project = await getProject(id)
  if (!project) return null
  return { ...project, pages: await listPages(id) }
}

export function projectStats(pages) {
  return {
    total: pages.length,
    processed: pages.filter(page => page.processedImage).length,
    transcribed: pages.filter(page => page.rawOcrText).length,
    reviewed: pages.filter(page => page.status === 'reviewed').length,
    needsReview: pages.filter(page => page.status === 'needs_review').length,
  }
}

export function missingPageNumbers(pages) {
  const numbers = [...new Set(pages.map(page => page.pageNumber))].sort((a, b) => a - b)
  if (numbers.length < 2) return []
  return Array.from({ length: numbers.at(-1) - numbers[0] + 1 }, (_, i) => numbers[0] + i).filter(n => !numbers.includes(n))
}

export async function touchProject(projectId) {
  const project = await getProject(projectId)
  if (project) await saveProject(project)
}

